### PR TITLE
[GUI] Better event filtering system

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #... (if any)

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

We used to support filter by only `type`.
Now we also support filter by `key`, or by `(type, key)` combination.

by key:
```py
import taichi as ti
gui = ti.GUI('Hello')
while not gui.get_event(ti.GUI.ESCAPE):  # has ESC event
    gui.show()
```

by key + type:
```py
import taichi as ti
gui = ti.GUI('Hello')
while not gui.get_event(ti.GUI.RELEASE, ti.GUI.ESCAPE):  # has REL or ESC event
    gui.show()
```

```py
import taichi as ti
gui = ti.GUI('Hello')
while not gui.get_event((ti.GUI.RELEASE, ti.GUI.ESCAPE)):  # has REL and ESC event
    gui.show()
```

by comb:
```py
import taichi as ti
gui = ti.GUI('Hello')
filt = [(ti.GUI.PRESS, ti.GUI.ESCAPE), (ti.GUI.RELEASE, ti.GUI.SPACE)]
while not gui.get_event(*filt):  # PRESS and ESC  or  RELEASE and SPACE
    gui.show()
```

### CODE REVIEW

**Comment by archibate** [comment]
TODO: yield, get_events

**Comment by archibate** [comment]
Hello?

File: python/taichi/misc/gui.py [hunk 1]
```diff
@@ -123,30 +123,53 @@ def line(self, begin, end, radius, color):
...
     def get_event(self, *filter):
+        types = []
+        keys = []
+        combs = []
+        for ent in filter:
+            if ent in [GUI.PRESS, GUI.RELEASE]:
+                types.append(ent)
+            elif isinstance(ent, tuple) or isinstance(ent, list):
+                type, key = ent
+                combs.append((type, key))
```
Comment (k-ye) [inline python/taichi/misc/gui.py:144]: Just `combs.append(tuple(ent))`?

File: python/taichi/misc/gui.py [hunk 2]
```diff
@@ -123,30 +123,53 @@ def line(self, begin, end, radius, color):
...
+            self.core.screenshot(file_or_image)
+        elif file_or_image is not None:
+            self.set_image(file_or_image)
         self.clear(self.background_color)
 
-    def has_key_event(self):
-        return self.core.has_key_event()
 
     def get_event(self, *filter):
+        types = []
```
Comment (k-ye) [inline python/taichi/misc/gui.py:136]: A bit more efficient if `types, keys, combs` are `set()`?

File: python/taichi/misc/gui.py [hunk 3]
```diff
@@ -123,30 +123,53 @@ def line(self, begin, end, radius, color):
...
+                type, key = ent
+                combs.append((type, key))
+            else:
+                keys.append(ent)
         while True:
             if not self.has_key_event():
                 return False
             self.event = self.get_key_event()
-            if not len(filter) or self.event.type in filter:
+            if not len(combs):
```
Comment (k-ye) [inline python/taichi/misc/gui.py:151]: I find it a bit hard to follow. I think the logic here can be broke down into a few simpler checks:

* if `types, keys, combs` are all empty: return `True`, because there's no filter
* Check `combs`
* Check `keys`
* Check `types`

wdyt?

File: python/taichi/misc/gui.py [hunk 4]
```diff
@@ -123,30 +123,53 @@ def line(self, begin, end, radius, color):
         self.canvas.path_single(begin[0], begin[1], end[0], end[1], color,
                                 radius)
 
-    def show(self, file=None):
+    def show(self, file_or_image=None):
```
Comment (k-ye) [inline python/taichi/misc/gui.py:126]: how about having two args: `file` and `image`? Then users can both show it and save it as files

File: python/taichi/misc/gui.py [hunk 5]
```diff
@@ -123,30 +123,53 @@ def line(self, begin, end, radius, color):
...
+            if not len(combs):
+                if not len(types) or self.event.type in types:
+                    if not len(keys) or self.event.key in keys:
+                        break
+            elif (self.event.type, self.event.key) in combs:
                 break
         return True
 
-    def get_key_event(self):
+    def get_cursor_pos(self):  # ABi
```
Comment (k-ye) [inline python/taichi/misc/gui.py:159]: What's `ABi`?

File: python/taichi/misc/gui.py [hunk 6]
```diff
@@ -158,6 +181,12 @@ def get_key_event(self):
         self.core.pop_key_event_head()
         return e
 
+    def peek_key_event(self):
+        e = None
```
Comment (k-ye) [inline python/taichi/misc/gui.py:185]: I'm afraid this is not peeking. Functions named `peek` usually implies idempotency, yet  inside `get_key_event()`, the event is popped.

**Comment by k-ye** [comment]
Sorry for the late response..

File: python/taichi/misc/gui.py (hunk 1 above)
Comment (archibate) [reply python/taichi/misc/gui.py:144]: We want to make sure that ent is a `(type, key)` instead of `(233, 666, 888)`.

File: python/taichi/misc/gui.py (hunk 5 above)
Comment (archibate) [reply python/taichi/misc/gui.py:159]: `get_cursor_pos` is an alias of `self.core.get_cursor_pos`, which is in C++ scope, so it's an ABI.

File: python/taichi/misc/gui.py (hunk 6 above)
Comment (archibate) [reply python/taichi/misc/gui.py:185]: How about
```
get_key_event -> pop_key_event
peek_key_event -> get_key_event
```
?

File: python/taichi/misc/gui.py (hunk 4 above)
Comment (archibate) [reply python/taichi/misc/gui.py:126]: I would suggest them:
1. `gui.set_image(image); gui.show(file)`.
2. `gui.show(); ti.imwrite(image, file)`.
I like 2.

File: python/taichi/misc/gui.py (hunk 5 above)
Comment (k-ye) [reply python/taichi/misc/gui.py:170]: hmm, feels a bit redundant, but sounds good

File: python/taichi/misc/gui.py (hunk 6 above)
Comment (k-ye) [reply python/taichi/misc/gui.py:196]: i think the current `get_key_event` is fine. Do we need peek now? If we do, maybe `get_xx()` should be implemented on top of `peek_xx()`..

**Comment by archibate** [comment]
Btw, do we have some deprecated function need to be marked as deprecated? Like `ti.x86_64 -> ti.x64`, `ti.global_var -> ti.var`? I found this https://python-deprecated.readthedocs.io/en/latest

File: python/taichi/misc/gui.py (hunk 4 above)
Comment (k-ye) [reply python/taichi/misc/gui.py:126]: Then i think this should just be `image`, because `imwrite` is already able to save to files? Again, if you think at the API layer, GUI is supposed to deal with one thing -- GUI. And we have separate APIs for file writings. WDYT?

File: python/taichi/misc/gui.py (hunk 6 above)
Comment (k-ye) [reply python/taichi/misc/gui.py:204]: Sorry for being picky here, but i see little value of having both `pop` and `get` -- According to the current implementation, `pop_key_event()` first `wait_key_event()`, meaning that `has_key_event()` in L205 is redundant. Again, I'd suggest we just keep `get_xx()` as is, it makes the API simpler.

**Comment by archibate** [comment]
Ok, now the API becomes:
peek: if `has` then `return e`, else `return None`.
get: if `has` then `pop` and `return e`, else `return None`.
wait: block until `has`.
wait_get: if `has` then `pop` and `return e`, else block until `has`, then `pop` and `return e`.

**Comment by archibate** [comment]
> Then i think this should just be image, because imwrite is already able to save to files? Again, if you think at the API layer, GUI is supposed to deal with one thing -- GUI. And we have separate APIs for file writings. WDYT?

1. our imread/imwrite seems to has some bug now (#802).
2. we need to ensure backward compatibility and make old examples work.
3. It's very convenient to screenshot in one line.
4. It's not a big deal to do an `isinstance`.

**Comment by k-ye** [comment]
I'll just leave this to @yuanming-hu  to decide the API changes. In general, I think there are a few things.

* This PR is, once again, getting too big and losing focus from what it intends to be. I thought you were just going to enhance the filtering, but now we have to debate over peek/get/wait_get. And it's pulling in non GUI event things like `gui.show` and `rgb_to_hex`.
* Again, while I think it's valuable to have both peek and get, this combination of `event`, `events`, `{get_, is_} x {key_pressed, is_pressed, key_event}` is just making things really confusing. Personally i think it would be better if we could first list out what APIs we want, then start implementing them.

**Comment by archibate** [comment]
Yes, it did get confusing... let's just merge the first commit in this PR?

**Comment by archibate** [comment]
I still don't want to break down this PR, since the lesson from GL PRs is still 历历在目... We shouldn't get so far at the first place: I added some 顺手牵羊 codes. 你没有阻止我那样做，甚至还对那些顺手牵羊的代码做了一些改进建议...

**Comment by yuanming-hu** [comment]
Sorry I have meetings until early afternoon. I'll take care of this thread later today. Thanks for being patient and for the valuable discussions here.

**Comment by archibate** [comment]
> 'm following this tutorial http://www.realtimerendering.com/raytracing/Ray%20Tracing%20in%20a%20Weekend.pdf and writing a toy ray tracer on Metal these days...

Oh this tutorial is cool! Have fun!

> (For example, I just found out that Metal doesn't support recursion function calls!)

Interesting, so does GLSL... Seems most morden shader langs don't support recursion, how about CUDA? And that may explain why most ray marching tutorials use CPU...
How do you deal with this limitation then? Maybe helpful in taichi's force-inlined function system too.

File: python/taichi/misc/gui.py (hunk 5 above)
Comment (yuanming-hu) [reply python/taichi/misc/gui.py:177]: Sorry but I'm still confused about this: does `ABi` here stand for [Application binary interface](https://en.wikipedia.org/wiki/Application_binary_interface)? Why is `i` lowercased here?

File: python/taichi/misc/gui.py [hunk 7]
```diff
@@ -155,10 +195,33 @@ def get_key_event(self):
...
+    def wait_key_event(self):  # ABi
+        self.core.wait_key_event()
+
+    def peek_key_event(self):  # Unique APi
+        e = None
+        if self.has_key_event():
+            e = self._peek_key_event()
+        return e
+
+    def get_key_event(self):  # Unique APi
```
Comment (yuanming-hu) [inline python/taichi/misc/gui.py:214]: Again, if `APi` stands for *application programming interfaces*, why is `i` lowercased?

File: python/taichi/misc/gui.py (hunk 4 above)
Comment (yuanming-hu) [reply python/taichi/misc/gui.py:126]: First of all, **this part does not belong to the event filtering system. Please revert**.

Please understand that having too many things in a PR makes reviewing really hard since the discussions cannot be focused on the real thing. This also increases the delay in merging your contribution since there are too many to discuss, and will result in nasty merge conflicts.


My thoughts on this specific issue: I agree that `ti.GUI.show` should not provide the option to write to disk. We might still need an API to get the image from the GUI or save the GUI image to disk, since sometimes people use drawing commands such as `ti.GUI.circles` to paint, and the painted result is only available from the GUI. Maybe we should do
 - `gui.set_image` to set a background image and allow people to paint on it
 - `gui.show(image=None)` to show the current canvas, or a plain 2D image (if `image` not none). Clear the temporary canvas after window content is updated.
 - `gui.screenshot(filename=None) -> np.ndarray` return the canvas. If `filename` is not `None`, use `imwrite` to save to disk. `ti.imwrite(gui.screenshot(), 'X.png')` sounds slightly verbose so I think a shortcut would be good.

The API design here should be put into a separate issue.

File: python/taichi/misc/gui.py (hunk 7 above)
Comment (archibate) [reply python/taichi/misc/gui.py:211]: For fun :)

**Comment by archibate** [comment]
Fully reverted, only event filtering preserved!

File: python/taichi/misc/gui.py [hunk 8]
```diff
@@ -139,24 +139,57 @@ def show(self, file=None):
...
+                elif ent in [GUI.PRESS, GUI.RELEASE]:
+                    self.types.add(ent)
+                else:
+                    self.keys.add(ent)
+
+        def match(self, e):
+            if (e.type, e.key) in self.combs:
+                return True
+            if not self.keys or e.key in self.keys:
+                if not self.combs or e.key in self.combs:
```
Comment (k-ye) [inline python/taichi/misc/gui.py:160]: Do you mean to check `types`? Also, it appears to me that L159 and L160 has an OR relationship, so i'd still suggest  to write flattened if checks.

File: python/taichi/misc/gui.py [hunk 9]
```diff
@@ -139,24 +139,57 @@ def show(self, file=None):
...
+            if filter is None or filter.match(e):
+                yield e
 
     def get_key_event(self):
         self.core.wait_key_event()
         e = GUI.Event()
         e.key = self.core.get_key_event_head_key()
         e.type = self.core.get_key_event_head_type()
         e.pos = self.core.get_key_event_head_pos()
+        e.pos = (e.pos[0], e.pos[1])
```
Comment (k-ye) [inline python/taichi/misc/gui.py:192]: nit: I think it's fine just to write `tuple(self.core.get_key_event_head_pos())`, without worrying  too much about the returned size. Or maybe `x, y = self.core.get_key_event_head_pos(); e.pos = (x, y)`?

File: python/taichi/misc/gui.py [hunk 10]
```diff
@@ -139,24 +139,57 @@ def show(self, file=None):
...
         return self.core.has_key_event()
 
     def get_event(self, *filter):
+        for e in self.get_events(*filter):
+            self.event = e
+            return True
+        return False
+
+    def get_events(self, *filter):
+        if not filter:
```
Comment (k-ye) [inline python/taichi/misc/gui.py:174]: nit: `filter = filter or None`?

Comment (archibate) [reply python/taichi/misc/gui.py:174]: Cool!

File: python/taichi/misc/gui.py (hunk 8 above)
Comment (archibate) [reply python/taichi/misc/gui.py:160]: > `types`

Nice catch! I couldn't imagine how my brain was fried to write such buggy code...

> Also

We want AND, see PR description, so that user can `get_event(ti.GUI.ESCAPE, ti.GUI.PRESS)` to filter only ESC press.

**Comment by archibate** [comment]
AppVeyor failed due to network reason again:
```
Build started
git clone -q https://github.com/taichi-dev/taichi.git C:\taichi
git fetch -q origin +refs/pull/801/merge:
git checkout -qf FETCH_HEAD
set TAICHI_REPO_DIR=C:\taichi
%PYTHON% %TAICHI_REPO_DIR%/misc/appveyor_filter.py || appveyor exit 0
cd C:\
curl --retry 10 --retry-delay 5 https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-msvc2017.zip -LO
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:34 --:--:--     0
curl: (52) Empty reply from server
Command exited with code 52
```

File: python/taichi/misc/gui.py (hunk 9 above)
Comment (k-ye) [reply python/taichi/misc/gui.py:189]: Ack

File: python/taichi/misc/gui.py (hunk 8 above)
Comment (k-ye) [reply python/taichi/misc/gui.py]: Then it seems that we have a discrepancy here:

1. According to your `by comb` example

```py
filt = [(ti.GUI.PRESS, ti.GUI.ESCAPE), (ti.GUI.RELEASE, ti.GUI.SPACE)]
while not gui.get_event(*filt):  # PRESS and ESC  or  RELEASE and SPACE
```

The elements in `get_event()`'s `*filter` argument have an OR relationship.

Yet in the `by key + type` example:

```py
while not gui.get_event(ti.GUI.RELEASE, ti.GUI.ESCAPE):  # has REL and ESC event
    gui.show()
```

The elements in `get_event()`'s `*filters` now have an AND relationship.

---

What do you think if we make this consistent by doing this:

* elements in the same tuple have an AND relationship
* elements in the top-level `*filter` args always have an OR relationship

Then for the `by comb` example, these are equivalent:

```py
gui.get_event((ti.GUI.PRESS, ti.GUI.ESCAPE), (ti.GUI.RELEASE, ti.GUI.SPACE))
gui.get_event(*[(ti.GUI.PRESS, ti.GUI.ESCAPE), (ti.GUI.RELEASE, ti.GUI.SPACE)])
```

And for the `by key + type` example, it becomes:

```py
gui.get_event((ti.GUI.RELEASE, ti.GUI.ESCAPE))
```

Then, the following two become equivalent and mean either `ti.GUI.RELEASE` or `ti.GUI.ESCAPE`, although the second one is unnecessarily verbose:

```
gui.get_event(ti.GUI.RELEASE, ti.GUI.ESCAPE)
gui.get_event((ti.GUI.RELEASE, ), (ti.GUI.ESCAPE, ))
```
---

FYI, I should have mentioned this earlier, but please leave the work of "marking comments as resolved" to the reviewers. Otherwise it's easy that we might miss some comments in the previous reviews.

Comment (archibate) [reply python/taichi/misc/gui.py]: Thanks, resolved now.

File: python/taichi/misc/gui.py [hunk 11]
```diff
@@ -139,24 +139,45 @@ def show(self, file=None):
...
         return self.core.has_key_event()
 
     def get_event(self, *filter):
+        for e in self.get_events(*filter):
+            self.event = e
+            return True
+        return False
+
+    def get_events(self, *filter):
+        filter = filter and GUI.EventFilter(*filter) or None
```
Comment (k-ye) [inline python/taichi/misc/gui.py:165]: I know I suggested `filter or None`, but in this case, i think the idiomatic way is `GUI.EventFilter(*filter) if filter else None`.

**Comment by archibate** [comment]
> 1.Since currently we don't check the tuple size, in theory a user could pass in an AND tuple with a size larger than 2. E.g. `(TI.RELEASE, TI.ESCAPE, TI.SPACE)`. Should we disallow this? Or should we consider supporting it in the future?
> 2. If we do, then are `(TI.RELEASE, TI.ESCAPE, TI.SPACE)` and `(TI.RELEASE, SPACE, TI.ESCAPE)` equivalent, or even `(TI.SPACE, TI.RELEASE, TI.ESCAPE)`?

Thank for purposing this, but we won't support this. Why can a event be both ESC and Space? There is only one key pre-event, no multiple keys, if they want, they use `gui.is_pressed(ESCAPE, SPACE)`. Let's just `assert len(the_tuple) == 2` now.

**Comment by archibate** [comment]
Postponed by #839.

**Comment by archibate** [comment]
added documention for gui.

File: docs/gui.rst [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+==========
+
+Taichi has a built-in GUI system to help users display graphic results easier.
+
+
+Create a window
+---------------
+
+
+.. function:: ti.GUI(title, res, bgcolor = 0x000000):
```
Comment (archibate) [inline docs/gui.rst:13]:
```suggestion
.. function:: ti.GUI(title, res, bgcolor = 0x000000)
```
